File: solutions/cpp/Copying_postings_list.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <random>

#include "./Postings_list_prototype.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::make_shared;
using std::random_device;
using std::shared_ptr;
using std::uniform_int_distribution;

// @include
shared_ptr<ListNode<int>> CopyPostingsList(
    const shared_ptr<ListNode<int>>& L) {
  // Returns empty list if L is nullptr.
  if (!L) {
    return nullptr;
  }

  // 1st stage: Copies the nodes from L.
  shared_ptr<ListNode<int>> p = L;
  while (p) {
    auto temp =
        make_shared<ListNode<int>>(ListNode<int>{p->data, p->next, nullptr});
    p->next = temp;
    p = temp->next;
  }

  // 2nd stage: Updates the jump field.
  p = L;
  while (p) {
    if (p->jump) {
      p->next->jump = p->jump->next;
    }
    p = p->next->next;
  }

  // 3rd stage: Restores the next field.
  p = L;
  shared_ptr<ListNode<int>> copied = p->next;
  while (p->next) {
    shared_ptr<ListNode<int>> temp = p->next;
    p->next = temp->next;
    p = temp;
  }
  return copied;
}
// @exclude

template <typename T>
void CheckPostingsListEqual(shared_ptr<ListNode<T>> a,
                            shared_ptr<ListNode<T>> b) {
  while (a && b) {
    cout << a->data << ' ';
    assert(a->data == b->data);
    assert((a->jump == shared_ptr<ListNode<T>>(nullptr) &&
            b->jump == shared_ptr<ListNode<T>>(nullptr)) ||
           (a->jump && b->jump && a->jump->data == b->jump->data));
    if (a->jump) {
      cout << a->jump->data;
    }
    cout << endl;
    a = a->next, b = b->next;
  }
  assert(a == shared_ptr<ListNode<T>>(nullptr) &&
         b == shared_ptr<ListNode<T>>(nullptr));
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> n_dis(1, 1000);
      n = n_dis(gen);
    }
    shared_ptr<ListNode<int>> L = nullptr;
    shared_ptr<ListNode<int>> curr = L;
    for (int i = 0; i < n; ++i) {
      auto temp = make_shared<ListNode<int>>(ListNode<int>{i, nullptr});
      if (L) {
        curr->next = temp;
        curr = temp;
      } else {
        curr = L = temp;
      }
      // Randomly assigned a jump node.
      uniform_int_distribution<int> dis(0, i + 1);
      int jump_num = dis(gen);
      shared_ptr<ListNode<int>> jump = L;
      while (jump_num--) {
        jump = jump->next;
      }
      temp->jump = jump;
    }
    shared_ptr<ListNode<int>> copied = CopyPostingsList(L);
    CheckPostingsListEqual<int>(L, copied);
  }
  return 0;
}

File: solutions/python/copying_postings_list.py
import sys
import random
from postings_list_prototype import ListNode


# @include
def copy_postings_list(L):
    if not L:
        return None

    # Stage 1: Makes a copy of the original list without assigning the jump
    #          field, and creates the mapping for each node in the original
    #          list to the copied list.
    it = L
    while it:
        new_node = ListNode(it.order, it.next, None)
        it.next = new_node
        it = new_node.next

    # Stage 2: Assigns the jump field in the copied list.
    it = L
    while it:
        if it.jump:
            it.next.jump = it.jump.next
        it = it.next.next

    # Stage 3: Reverts the original list, and assigns the next field of
    #          the copied list.
    it = L
    new_list_head = it.next
    while it.next:
        it.next, it = it.next.next, it.next
    return new_list_head
# @exclude


def check_postings_list_equal(a, b):
    while a and b:
        print(a.order, end=' ')
        assert a.order == b.order
        assert ((a.jump is None and b.jump is None) or
                (a.jump and b.jump and a.jump.order == b.jump.order))
        if a.jump:
            print(a.jump.order, end='')
        print()
        a = a.next
        b = b.next
    assert a is None and b is None


def main():
    for _ in range(1000):
        if len(sys.argv) == 2:
            n = int(sys.argv[1])
        else:
            n = random.randint(1, 1000)
        L = curr = None
        for i in range(n):
            temp = ListNode(i, None)
            if L:
                curr.next = temp
                curr = temp
            else:
                curr = L = temp
            # Randomly assigned a jump node.
            jump_num = random.randint(0, i + 1)
            jump = L
            while jump_num:
                jump_num -= 1
                jump = jump.next
            temp.jump = jump

        copied = copy_postings_list(L)
        check_postings_list_equal(L, copied)


if __name__ == '__main__':
    main()
